Fixed duplicate services and constistency in naming

Meziu opened the pull request:
Weirdly, we missed one of the biggest memory safety issues in this whole library: there may be more than one handle to a service at once, yet services would break if more than an instance existed at a time.

For example:
```Rust
let gfx = Gfx::default();
let gfx2 = Gfx::default();

drop(gfx2);

// Now `gfx` is completely invalid and the program will crash when dropping it 
// since it'll try to double-free memory
```

Some services (like `Gfx`) completely crash the program, others simply stop working, but all are affected in an unexpected behaviour.


I wanted to keep the "own the service" behaviour of having a struct to hold them, because it fits nicely with Rust's `drop` mechanism, so the solution was to have an `AtomicBool` signaling the presence of the struct at runtime. This also comes with a new `ctru::Error` field to signal multiple instancing (we can add as many of those as we want since the `Os` field is separate). 

This also comes with a different `Gfx` default impl, since using the trait wouldn't work anymore with the possible error handling (changed `::default()` with `::init_default()` which now returns a `ctru::Result` object). At last, I've added tests in all service modules to test for these new changes. I believe we should start adding them from now on, since testing is an absolute breeze.

Overall it's a gigantic PR with many breaking changes, but it is absolutely needed something like this system to keep the whole library memory safe. Hope crashes won't be a thing anymore in the future.

AzureMarker commented:
Where is the double init problem documented? I haven't run across that before.

Meziu commented:
> Where is the double init problem documented? I haven't run across that before.

The issue isn’t with double initialisation, it’s with double ownership. As I showed in my snippet, if one of the two `Gfx` instances could get dropped before the other, but the drop code closes the service anyways, making the other struct invalid (functions would either do nothing or crash, depending on the service). We also can’t solve it with an `Rc` since that would undo the principle of mutability of the service’s state (Eg. keeping track of the current setup for the `Gfx` screens or `RomFS` mount).

Meziu commented:
> Weirdly, we missed one of the biggest memory safety issues in this whole library: there may only be more than one handle to a service at once, yet services would break if more than an instance existed at a time.
> 
> For example:
> 
> ```rust
> let gfx = Gfx::default();
> let gfx2 = Gfx::default();
> 
> drop(gfx2);
> 
> // Now `gfx` is completely invalid and the program will crash when dropping it 
> // since it'll try to double-free memory
> ```
> 
> Some services (like `Gfx`) completely crash the program, others simply stop working, but all are affected in an unexpected behaviour.
> 
> I wanted to keep the "own the service" behaviour of having a struct to hold them, because it fits nicely with Rust's `drop` mechanism, so the solution was to have an `AtomicBool` signaling the presence of the struct at runtime. This also comes with a new `ctru::Error` field to signal multiple instancing (we can add as many of those as we want since the `Os` field is separate).
> 
> This also comes with a different `Gfx` default impl, since using the trait wouldn't work anymore with the possible error handling (changed `::default()` with `::init_default()` which now returns a `ctru::Result` object). At last, I've added tests in all service modules to test for these new changes. I believe we should start adding them from now on, since testing is an absolute breeze.
> 
> Overall it's a gigantic PR with many breaking changes, but it is absolutely needed something like this system to keep the whole library memory safe. Hope crashes won't be a thing anymore in the future.

AzureMarker commented on ctru-rs/src/lib.rs:
```diff
@@ -38,12 +38,31 @@ pub fn init() {
+            // before the cleanup, though the time window would be almost nonexistent, and it would only
+            // really be a problem in preemptive threads.
+            //
+            // TL;DR : This code is bad.
```
Maybe we need to do something a little more fancy than an atomic boolean.

One idea is having a static `Mutex<usize>` which starts at zero and counts the number of open service handles. When it reaches zero again in drop, we can call the exit function. It needs to be a Mutex and not just an AtomicUsize because we need to hold the lock while we call the `hidExit` function, for example. Otherwise, we might decrement to zero, another thread calls `Hid::init` and increments to one (calling `hidInit`), and the drop function finally calls `hidExit`.

This would let us allow multiple handles to be created without issue, and only exiting the service once all are gone. Basically re-implementing reference counting, but it's a little hard to reuse `Arc` for this (feel free to try if you want, but it will probably be more verbose and complex).

We could code this pattern up into a `ServiceCleaner` or something (name tbd), like so (untested):
```rust
struct ServiceCleaner {
    counter: &'static Mutex<usize>
    exit: extern "C" fn()
}

impl ServiceCleaner {
    fn init(counter: &'static Mutex<usize>, exit: extern "C" fn()) -> Self {
        let mut value = counter.lock().unwrap() // todo: handle poisoning
        *value += 1;
        drop(value);
        Self { counter, exit }
    }
}

impl Drop for ServiceCleaner {
    fn drop(&mut self) {
        let mut value = self.counter.lock().unwrap(); // should probably handle poisoning - could just map_err to ignore it.
        *value -= 1;
        if *value == 0 {
            unsafe { self.exit() };
        }
    }
}

```

AzureMarker replied:
One place to be careful about with the above is `Gfx`. We should probably still only let them have one instance of that struct.

Meziu replied:
`Gfx` absolutely must not be re-initialised, since it takes parameters that change the internal state. Other services, like `Hid` or `SslC` have no problem being shared across threads, though we should be careful with what the services do.

AzureMarker replied:
Regarding the function naming, `init` is probably fine so long as the underlying ctru_sys funciton (ex. `aptInit`) is OK if it's called multiple times.

I checked all of the "*Init()" functions in libctru, and they all\* actually handle init/exit with reference counting correctly. So we could don't need to do anything special. Even the `gfxInit` function is partially guarded (the `gspInit` call only inits once, but the rest of the gfx settings get reset on every call).

\* romfsInit does try to mount it every time, so it might error out. And we still may still want to limit the Gfx initialization.

Meziu commented on ctru-rs/src/services/sslc.rs:
```diff
@@ -1,35 +1,58 @@
-                Err(r.into())
-            } else {
-                Ok(())
-            }
+        let r = unsafe { ctru_sys::sslcGenerateRandomData(buf.as_ptr() as _, buf.len() as u32) };
+        if r < 0 {
+            Err(r.into())
+        } else {
+            Ok(())
         }
     }
```
We talked about it when implementing `getrandom` in `linker-fix-3ds`. `SslC` looks to be just an overhead of `PS`, so we decided to use the “lower level” one.

ian-h-chamberlain commented on ctru-rs/src/error.rs:
```diff
@@ -28,6 +29,7 @@ impl fmt::Debug for Error {
                 .field("summary", &R_SUMMARY(err))
                 .field("level", &R_LEVEL(err))
                 .finish(),
+            Error::ServiceAlreadyActive(service) => write!(f, "Service {service} already active"),
```
We could use enum variants instead of the inner `&'static str` too if we wanted, since there is a finite well-known set of services. Might be annoying to keep in sync as we add new modules, though?

ian-h-chamberlain commented:
As an alternative design, I wonder if we could consider something like the singleton pattern laid out [in the embedded book](https://docs.rust-embedded.org/book/peripherals/singletons.html#how-do-we-do-this-in-rust).

I think it would look basically like
```rs
#[derive(Default)]
struct Services {
	gfx: Option<Gfx>,
	hid: Option<Hid>,
	//...
}

static mut SERVICES: Services = Services::default();
```

Then a user could either request a given service (`Gfx::try_init` or something maybe?), which could initialize and either create or replace the global `Option<Gfx>`. This would perhaps require APIs like `Gfx::force_init` or something if the user wanted to replace the existing service, but it seems like that might be necessary to support such a use case anyway.

There would probably still need to be some kind of lock on the singleton, for thread safety, but since it's `static` it should be safe to hand out references to the underlying service handles (`&'static Gfx` - is `&mut` okay here too?). We might want to avoid the `take_serial()` example they provide, since we wouldn't want to allow a user to call `take_gfx()` followed by `gfx_init` or something and end up with two "live" `Gfx`... but I think in principle this could work to ensure services are only initialized one at a time.

AzureMarker commented:
That singleton pattern is interesting (I haven't seen it before), but since libctru already reference counts most of the services, I think we should keep that pattern.

Meziu commented:
Fixed the nits and such. I believe the `ServiceCloser` idea would be better, but I won't be able to thoroughly test it for the time being.

Meziu commented:
Oh well, trying to test things out made me realize an issue with the `ServiceCloser` idea: `Mutex`es aren't `const`.
The idea of keeping the lock is good, but it's a problem to implement...

AzureMarker commented:
We can use the Mutex from the `spin` crate like we do with `RwLock` in pthread-3ds.

AzureMarker commented on ctru-rs/src/gfx.rs:
```diff
@@ -227,6 +227,9 @@ mod tests {
     fn gfx_duplicate() {
         // We don't need to build a `Gfx` because the test runner has one already
-        assert!(Gfx::init_default().is_err());
+        match Gfx::init() {
+            Err(Error::ServiceAlreadyActive("Gfx")) => return,
+            _ => panic!(),
+        }
```
nit: this could be:
```suggestion
        assert!(matches!(Gfx::init(), Err(Error::ServiceAlreadyActive("Gfx"))));
```

AzureMarker commented on ctru-rs/src/error.rs:
```diff
@@ -29,7 +29,10 @@ impl fmt::Debug for Error {
-            Error::ServiceAlreadyActive(service) => write!(f, "Service {service} already active"),
+            Error::ServiceAlreadyActive(service) => f
+                .debug_tuple("ServiceAlreadyActive")
+                .field(&String::from(service))
```
nit: this could just be the following?
```suggestion
                .field(service)
```

AzureMarker commented:
We may also be able to avoid any reference counting on our end for most services based on this:
https://github.com/Meziu/ctru-rs/pull/50#discussion_r817347456

Meziu commented:
> We can use the Mutex from the `spin` crate like we do with `RwLock` in pthread-3ds (yielding instead of busy-looping).

I went with `SyncLazy`. I believe it's better because it only initializes when needed, and not in all programs.

Meziu commented:
Ok, I changed most of everything. Now it uses `Mutex`, and all services have an abstract implementation of `ServiceHandler` (did it this way because the `init` functions had to be in the lock's lifetime too, and also because it was easy to expand). This is nowhere near final.

@AzureMarker there are many things you'll have to review, though one of the smallest is the presence of `Error::ServiceAlreadyActive`. With this implementation, it's used only by `Gfx`, so I removed the `&str` fields altogether (they were also really ugly).

AzureMarker commented on ctru-rs/src/services/mod.rs:
```diff
@@ -9,3 +9,48 @@ pub mod sslc;
+
+use crate::Error;
+use std::sync::Mutex;
+pub(crate) struct ServiceHandler {
```
nit: this should go in a separate file.

AzureMarker commented on ctru-rs/src/lib.rs:
```diff
@@ -3,6 +3,7 @@
 #![feature(test)]
 #![feature(custom_test_frameworks)]
 #![test_runner(test_runner::run)]
+#![feature(once_cell)]
```
nit: we can use the once_cell crate instead to avoid adding another feature here.

AzureMarker commented on ctru-rs/src/lib.rs:
```diff
@@ -28,6 +29,7 @@ pub fn init() {
 
     use std::panic::PanicInfo;
 
+    #[cfg(not(test))]
```
nit: this panic hook setup could be in a separate function so we only need one `cfg(not(test))` and it doesn't get interwoven in `init`.

AzureMarker commented on ctru-rs/src/romfs.rs:
```diff
@@ -54,10 +54,14 @@ mod tests {
+
+        assert_eq!(lock, 1);
+
+        drop(_romfs);
```
Does this not cause a deadlock? Wouldn't the first lock not get dropped until the end of scope?

Meziu replied:
Yeah I named the variable wrongly. It doesn’t hold the lock (eve  though it’s called `lock`), it just holds the value by using a temporary lock.

AzureMarker commented on ctru-rs/src/services/apt.rs:
```diff
@@ -1,25 +1,38 @@
 #[non_exhaustive]
-pub struct Apt(());
+pub struct Apt {
+    _service_handler: ServiceHandler,
```
Most of these services don't need to use `ServiceHandler` because they are already reference counted by libctru:
https://github.com/devkitPro/libctru/blob/ebb53051886cfe966f1aa27ff3dd6689176de483/libctru/source/services/apt.c#L195
https://github.com/devkitPro/libctru/blob/ebb53051886cfe966f1aa27ff3dd6689176de483/libctru/source/services/apt.c#L345

Meziu replied:
But there isn’t an automatic drop, the decrement is still only available in the `aptExit`.

AzureMarker replied:
We can call the exit functions in our drop like the code in master does. The service will only actually exit once its internal reference count is zero.

AzureMarker commented on ctru-rs/src/services/apt.rs:
```diff
@@ -1,25 +1,38 @@
-        }
+
+                Ok(())
+            },
+            // `socExit` returns an error code. There is no documentantion of when errors could happen,
+            // but we wouldn't be able to handle them in the `Drop` implementation anyways.
+            // Surely nothing bad will happens :D
+            || unsafe {
+                // The socket buffer is freed automatically by `socExit`
+                ctru_sys::aptExit();
```
The SOC comments got copied around too much and aren't relevant here.

Meziu replied:
Ooops, yeah…

AzureMarker commented on ctru-rs/src/romfs.rs:
```diff
@@ -11,39 +11,39 @@
-            Ok(_) => {
+        let _service_handler = ServiceHandler::new(
+            &ROMFS_ACTIVE,
+            true,
```
We should probably set this to false.

Meziu replied:
Why? The start code isn’t run again.

AzureMarker replied:
That's fair. I guess if we add `&mut self` functions we should switch it to false, but it's fine for now.

AzureMarker commented on ctru-rs/src/services/mod.rs:
```diff
@@ -9,3 +9,48 @@ pub mod sslc;
+
+use crate::Error;
+use std::sync::Mutex;
+pub(crate) struct ServiceHandler {
```
nit: Using `Handler` in a type name usually means there is a better name that can be used instead. `ServiceCloser` or `ServiceReference` are some ideas.

Meziu replied:
`ServiceCloser` is misleading, since it also starts the service.

AzureMarker commented on ctru-rs/src/services/soc.rs:
```diff
@@ -1,18 +1,18 @@
 /// such as those found in `std::net`. The service will be closed when this struct is is dropped.
 #[non_exhaustive]
 pub struct Soc {
-    soc_mem: *mut u32,
```
We should still deallocate the memory. To store the memory address, you might need another Mutex.

AzureMarker replied:
nvm, turns out we lose access to this memory when we pass it in.

AzureMarker commented on ctru-rs/src/services/soc.rs:
```diff
@@ -31,20 +31,28 @@ impl Soc {
+
+                Ok(())
+            },
+            // `socExit` returns an error code. There is no documentantion of when errors could happen,
+            // but we wouldn't be able to handle them in the `Drop` implementation anyways.
+            // Surely nothing bad will happens :D
```
We could panic in the drop, but it seems a little over the top.

Meziu replied:
I’d avoid panicking. Also, if the code works properly (and thus the service is actually open when calling the `Drop`) then I believe there aren’t any problems.

AzureMarker commented on ctru-rs/src/services/soc.rs:
```diff
@@ -31,20 +31,28 @@ impl Soc {
+            // but we wouldn't be able to handle them in the `Drop` implementation anyways.
+            // Surely nothing bad will happens :D
+            || unsafe {
+                // The socket buffer is freed automatically by `socExit`
+                ctru_sys::socExit();
```
Is this true? I couldn't find the place it's freed in libctru. Also, does that mean that the previous code was doing a double free?

Meziu replied:
While searching around for this buffer, I saw this note in the docs:

> Note: The specified context buffer can no longer be accessed by the process which called this function, since the userland permissions for this block are set to no-access.

Asking in the homebrew forums for confirmation, Fenrir himself responded telling me a free is probably the case, possibly handled by the OS itself when the process is closed. A double free panic didn’t happen possibly because the functions (`free` has no way of error handling) simply ignore OS errors.

AzureMarker commented on ctru-rs/src/lib.rs:
```diff
@@ -27,9 +26,13 @@ pub fn init() {
         libc::atexit(services_deinit);
     }
 
+    #[cfg(test)]
+    panic_hook_setup();
```
This should be `cfg(not(test))`

AzureMarker commented on ctru-rs/src/services/mod.rs:
```diff
@@ -5,52 +5,10 @@ pub mod hid;
 pub mod ps;
 pub mod soc;
 pub mod sslc;
+mod reference;
```
I think you forgot to commit `reference.rs`

Meziu commented:
I reverted all services not needing the refcount (`Soc` doesn't have ref-count by itself, nor `RomFS`).

AzureMarker commented on ctru-rs/src/lib.rs:
```diff
@@ -26,6 +26,12 @@ pub fn init() {
+    _panic_hook_setup();
+}
+
+fn _panic_hook_setup() {
```
nit: This shouldn't be prefixed with `_`. It's already a private function because it doesn't have `pub`.

AzureMarker commented on ctru-rs/src/lib.rs:
```diff
@@ -38,11 +44,13 @@ pub fn init() {
             println!("\nPress SELECT to exit the software");
-            let hid = services::hid::Hid::init().unwrap();
+
+            let hid = Hid::init().unwrap();
```
A little out of scope for this PR, but maybe we should not unwrap here? Instead if there's an error, just return?

I don't have much opinion either way, but it stood out to me.

Meziu replied:
Yeah, you’re right. Maybe I could print an error if the user is catching `stdout`, but panicking is a bad idea.

AzureMarker commented on ctru-rs/src/services/reference.rs:
```diff
@@ -0,0 +1,44 @@
+        S: FnOnce() -> crate::Result<()>,
+        E: Fn() + 'static,
+    {
+        let mut value = counter.lock().unwrap(); // todo: handle poisoning
```
Do you want to handle the poisoning TODOs in this PR?

Meziu replied:
I’d like to. What can we do about it, though?

AzureMarker replied:
We can either ignore it, unwrap/panic, or handle it. In our case, what it means to be poisoned is that the lock may have been held during a panic, which dropped the lock. So, maybe the start function panics, for example. Or, start doesn't panic, but somehow there's a panic between start and the counter increment.

If we ignore it, that is probably fine, but we just need to make sure we can't panic between the start task succeeding and the counter getting incremented. This is probably fine. Unwrapping/panicking is also an option, and would be the safest, plus it's unlikely that the lock gets poisoned in the first place, so this is fine to do. Handling the poison error seems hard because that implies fixing the broken state. But it's hard to know what's broken. So not much to do here I think.

I may be missing a few things since this is all off the top of my head.

Meziu replied:
I guess I'll change from an `unwrap` to an expect with a bit of info, but other than that it's hard. Still, panicking the thread while holding the lock is pretty much impossible, so I don't think there are issues.

AzureMarker commented on ctru-rs/src/services/reference.rs:
```diff
@@ -0,0 +1,44 @@
+use std::sync::Mutex;
+pub(crate) struct ServiceReference {
+    counter: &'static Mutex<usize>,
+    close: Box<dyn Fn()>,
```
I wonder if we should make this `dyn Fn() + Send + Sync`. I think this PR as-is would remove the `Send` and `Sync` auto-impls for the services that hold this struct.

AzureMarker commented on ctru-rs/src/romfs.rs:
```diff
@@ -10,27 +10,58 @@
+    use super::*;
+
+    #[test]
+    fn romfs_duplicate() {
```
nit: I think some of these test names are a little inaccurate. This one isn't testing duplicates, for instance.

AzureMarker reviewed (commented):
Just a few nits and questions. The PR looks good overall.

Meziu commented:
Changed `Soc`'s `allow_multiple` to false, because the init is handled with a parameter and we can't return an `Ok` without telling the user the change won't be made.
